File: app/presence/vessels.py
"""Vessels broadcasting AIS, live and unstored (#954).

Two requests to one open feed: where every vessel is, and what every vessel is.
The position message and the static message are separate on the wire and
separate here, joined by MMSI, because that is how AIS itself works — a vessel
transmits its position every few seconds and its name every few minutes, and a
vessel heard once may have one without the other.

**Coverage is receiver coverage, and it is coastal.** The feed is a national
authority's terrestrial receiver network. It sees its own sea area well and the
open ocean not at all, and no part of this layer may imply otherwise: an empty
Atlantic on this map is an empty receiver map, not an empty ocean. Satellite
AIS, which would see the rest, is a paid product and is not used.

Measured live on 2026-08-13: 1,261 positions and 1,203 static rows, every one
of the latter carrying a name. Ship type on all of them, which is why the
filter rows are read off the wire rather than inferred.
"""

from __future__ import annotations

import time
from datetime import UTC, datetime
from typing import Any

import httpx

from app.presence import vessels_no
from app.presence.registry import PresenceSource, source_for
from app.presence.vessel_types import category_for, nav_status_for

_TIMEOUT_S = 10.0
_USER_AGENT = "OSINT-console/1.0 (https://github.com/BasilSuhail/OSINT)"

#: A position older than this is not "now". Vessels are slow, so the tolerance
#: is wider than an aircraft's — a ship reports every few seconds under way but
#: only every few minutes at anchor, and dropping those would empty every
#: harbour on the map.
_MAX_POSITION_AGE_S = 60 * 60

_cache: dict[str, tuple[float, dict]] = {}


def _new_client() -> httpx.Client:
    """The outbound client, in one place so a test can replace it."""
    return httpx.Client(timeout=_TIMEOUT_S, headers={"User-Agent": _USER_AGENT})


def clear_cache() -> None:
    """Forget what is held in memory. Tests need this; nothing else does."""
    _cache.clear()


def _number(value: Any) -> float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    return float(value)


def _int(value: Any) -> int | None:
    if isinstance(value, bool) or not isinstance(value, int):
        return None
    return value


def _text(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    stripped = value.strip()
    return stripped or None


def _course(value: Any) -> float | None:
    """A course or heading, or nothing.

    AIS sends 511 for "heading not available" and 360 for "course not
    available". Drawn as a number, either would point a hull at a direction
    nobody transmitted.
    """
    number = _number(value)
    if number is None or number < 0 or number >= 360:
        return None
    return number


def normalise(feature: dict, static: dict | None, *, now_ms: float) -> dict | None:
    """One vessel, or None when there is nothing to draw.

    A position with no static row is still a vessel and is still drawn — it is
    really out there — with no name and no category. A static row with no
    position is dropped: there is nothing to put on a map.
    """
    if not isinstance(feature, dict):
        return None
    geometry = feature.get("geometry") or {}
    coordinates = geometry.get("coordinates")
    if not isinstance(coordinates, (list, tuple)) or len(coordinates) < 2:
        return None
    lon, lat = _number(coordinates[0]), _number(coordinates[1])
    if lat is None or lon is None:
        return None

    props = feature.get("properties") or {}
    reported_ms = _number(props.get("timestampExternal"))
    if reported_ms is not None and (now_ms - reported_ms) > _MAX_POSITION_AGE_S * 1000:
        return None

    mmsi = _int(props.get("mmsi")) or _int(feature.get("mmsi"))
    ship_type = _int((static or {}).get("shipType"))
    #: Zero is the standard's "not available", and reading it as a type would
    #: file a silent vessel under a category it never claimed.
    ship_type = ship_type or None

    return {
        "mmsi": mmsi,
        "name": _text((static or {}).get("name")),
        "callsign": _text((static or {}).get("callSign")),
        "imo": _int((static or {}).get("imo")) or None,
        "lat": lat,
        "lon": lon,
        "speed_kt": _number(props.get("sog")),
        "course": _course(props.get("cog")),
        "heading": _course(props.get("heading")),
        "nav_status": nav_status_for(_int(props.get("navStat"))),
        "category": category_for(ship_type),
        "ship_type": ship_type,
        "destination": _text((static or {}).get("destination")),
        "position_accurate": bool(props.get("posAcc")) if "posAcc" in props else None,
        #: Filled in by `flag_spoofed` once every vessel in the refresh is
        #: known: one of the two tests needs to see the others.
        "position_suspect": None,
        "reported_at": (
            datetime.fromtimestamp(reported_ms / 1000, UTC).replace(microsecond=0).isoformat()
            if reported_ms is not None
            else None
        ),
    }


#: No ship goes this fast. The fastest passenger craft in service run to about
#: 45 knots and everything else is far slower, so a hull reporting 46 knots is
#: not reporting where it is.
IMPLAUSIBLE_SPEED_KT = 40.0

#: Coordinates rounded to this many decimals — about a hundred metres — before
#: looking for a pile-up. Two vessels genuinely moored abreast are metres
#: apart; a spoofer puts many of them on one point.
_STACK_DECIMALS = 3

#: How many hulls on one point stop being a coincidence. Three is already
#: impossible for anything under way.
_STACK_MIN = 3

#: Only vessels making way are counted into a pile-up. Moored craft genuinely
#: do share a point — a row of pilot boats at one pier rounds to one set of
#: coordinates, and the first version of this test accused every marina in the
#: sea area. Hulls that are *moving* cannot occupy the same hundred metres.
_STACK_MOVING_KT = 5.0


def flag_spoofed(rows: list[dict]) -> list[dict]:
    """Mark positions the console does not believe, and say why.

    Not dropped, and this is deliberate. A transmitter claiming to be a ship
    in the middle of a forest is a real thing that is really happening, and it
    is worth more to a reader than the traffic around it — interference of
    this kind is a finding, not noise. What would be dishonest is drawing it
    like an ordinary vessel, so the mark says it is not trusted and the card
    says which test it failed.

    Two tests, both from what the feed itself sent:

    ``speed`` — a reported speed no vessel can reach.

    ``stacked`` — several hulls sharing one position to within about a hundred
    metres. Measured live on 2026-08-13: eight vessels on a single point far
    inland, reporting 35 to 51 knots, every one of them with the position
    accuracy flag set to false.
    """

    def moving(row: dict) -> bool:
        speed = row.get("speed_kt")
        return speed is not None and speed >= _STACK_MOVING_KT

    buckets: dict[tuple[float, float], int] = {}
    for row in rows:
        if not moving(row):
            continue
        key = (round(row["lat"], _STACK_DECIMALS), round(row["lon"], _STACK_DECIMALS))
        buckets[key] = buckets.get(key, 0) + 1

    for row in rows:
        key = (round(row["lat"], _STACK_DECIMALS), round(row["lon"], _STACK_DECIMALS))
        speed = row.get("speed_kt")
        #: Stacking is the stronger evidence, so it is the reason that shows.
        if moving(row) and buckets.get(key, 0) >= _STACK_MIN:
            row["position_suspect"] = "stacked"
        elif speed is not None and speed >= IMPLAUSIBLE_SPEED_KT:
            row["position_suspect"] = "speed"
        else:
            #: Set on every row, not only the suspect ones: a key that appears
            #: when there is bad news and is missing otherwise is a key every
            #: reader of this data has to remember to check for.
            row["position_suspect"] = None
    return rows


def merge(positions: dict, statics: list[dict], *, now_ms: float | None = None) -> list[dict]:
    """Positions, each carrying whatever static data arrived for the same MMSI.

    Deduped by MMSI: a feed that repeats a vessel is describing one vessel, and
    two marks for one hull would be counted twice by every reader looking at
    the map.
    """
    stamp = time.time() * 1000 if now_ms is None else now_ms
    by_mmsi: dict[int, dict] = {}
    for row in statics:
        key = _int(row.get("mmsi")) if isinstance(row, dict) else None
        if key is not None:
            by_mmsi[key] = row

    out: dict[int | None, dict] = {}
    features = positions.get("features") if isinstance(positions, dict) else None
    for feature in features or []:
        props = (feature or {}).get("properties") or {}
        key = _int(props.get("mmsi")) or _int((feature or {}).get("mmsi"))
        item = normalise(feature, by_mmsi.get(key) if key is not None else None, now_ms=stamp)
        if item is None:
            continue
        out[item["mmsi"]] = item
    return sorted(
        flag_spoofed(list(out.values())),
        key=lambda v: (v["mmsi"] is None, v["mmsi"] or 0),
    )


def _get_json(client: httpx.Client, source: PresenceSource, path: str) -> Any:
    """Try each endpoint in turn; the last failure is the one that surfaces."""
    last: Exception | None = None
    for base in source.endpoints:
        try:
            response = client.get(f"{base}{path}")
            response.raise_for_status()
            return response.json()
        except Exception as exc:
            last = exc
    raise last or LookupError("no endpoint configured")


def _fetch(client: httpx.Client, source: PresenceSource) -> dict:
    """Positions and static data, with partial failure kept partial.

    Losing the names is not losing the vessels. A refused static request leaves
    a map of unnamed marks, which is worse than a named one and far better than
    an empty one, and the response says `degraded` so the map can say so too.
    """
    degraded = False
    attributions: list[str] = []
    positions: dict = {}
    try:
        payload = _get_json(client, source, "/api/ais/v1/locations")
        positions = payload if isinstance(payload, dict) else {}
        attributions.append(source.attribution)
    except Exception:
        degraded = True

    statics: list[dict] = []
    try:
        payload = _get_json(client, source, "/api/ais/v1/vessels")
        statics = (
            [row for row in payload if isinstance(row, dict)] if isinstance(payload, list) else []
        )
    except Exception:
        degraded = True

    #: The second sea area, when it is configured (#954). Its rows are
    #: reshaped into this feed's shape on the way in, so the categories, the
    #: join, the spoofing tests and the age cut are written once and every
    #: source is held to them.
    #:
    #: No credentials is not degradation. The layer is not asking, which is a
    #: different thing from asking and being refused, and saying "degraded"
    #: for a source nobody configured would cry wolf on every console that
    #: only wants the first sea.
    if vessels_no.credentials() is not None:
        try:
            no_source = source_for("vessels_no")
            features, no_statics = vessels_no.fetch(
                client, no_source.endpoints[0], "/v1/latest/combined"
            )
            existing = positions.get("features") if isinstance(positions, dict) else None
            positions = {
                "type": "FeatureCollection",
                "features": [*(existing or []), *features],
            }
            statics = [*statics, *no_statics]
            if features:
                attributions.append(no_source.attribution)
        except Exception:
            degraded = True

    vessels = merge(positions, statics)
    return {
        "fetched_at": datetime.now(UTC).replace(microsecond=0).isoformat(),
        "count": len(vessels),
        "sources": attributions,
        "vessels": vessels,
        "degraded": degraded,
    }


def live_vessels(*, client: httpx.Client | None = None) -> dict:
    """The current picture, or an honest blank.

    When nothing answers, the result is empty and says so. Holding the last
    positions on screen would present a ship where it was an hour ago as where
    it is, which is the one thing a live layer must never do.
    """
    source = source_for("vessels")
    hit = _cache.get("vessels")
    now = time.monotonic()
    if hit is not None and now - hit[0] < source.ttl_s:
        return hit[1]

    owned = client is None
    http = client or _new_client()
    try:
        answer = _fetch(http, source)
    finally:
        if owned:
            http.close()

    #: An empty degraded picture is never cached: pinning "nothing" in place
    #: would turn one refused request into a minute of blank sea.
    if not (answer["degraded"] and answer["count"] == 0):
        _cache["vessels"] = (now, answer)
    return answer

File: osint-frontend/lib/vessels.ts
/** Vessels broadcasting AIS, live and unstored (#954).
 *
 * The category is transmitted, not inferred: AIS carries a ship type in the
 * static message every vessel sends, so "fishing" and "tankers" are separate
 * switches without a model or a guess behind them.
 *
 * Coverage is one authority's terrestrial receivers. This layer covers a sea
 * area, not an ocean, and the rail says so — an empty Atlantic on this map is
 * an empty receiver map.
 */

export type VesselCategory =
  | "cargo"
  | "tanker"
  | "passenger"
  | "fishing"
  | "pleasure"
  | "service"
  | "other"

export interface PresenceVessel {
  mmsi: number | null
  name: string | null
  callsign: string | null
  imo: number | null
  lat: number
  lon: number
  speed_kt: number | null
  /** Course over the ground. Null when the vessel sent the "not available"
   *  code, which must never be drawn as a direction. */
  course: number | null
  heading: number | null
  nav_status: string | null
  category: VesselCategory
  ship_type: number | null
  destination: string | null
  /** Whether the vessel said its own fix was accurate. */
  position_accurate: boolean | null
  /** Why this position is not believed, or null. `speed` — a speed no ship
   *  reaches. `stacked` — several moving hulls on one point. Both are the
   *  fingerprint of interference rather than of a vessel. */
  position_suspect: "speed" | "stacked" | null
  reported_at: string | null
}

export interface VesselAnswer {
  fetched_at: string
  count: number
  /** Attribution for every feed that answered this refresh. Rendered rather
   *  than hardcoded: the layer can cover more than one sea area, each with its
   *  own licence and its own notice to carry, and a console that names a
   *  source it did not use is as wrong as one that names none. */
  sources?: string[]
  vessels: PresenceVessel[]
  degraded: boolean
}

/** The notice to print under the layer. Falls back to nothing rather than to a
 *  guess: an unattributed feed is a licence problem, and inventing the line
 *  would hide it. */
export function vesselAttribution(sources: string[] | undefined): string | null {
  const named = (sources ?? []).filter((s) => s.trim())
  return named.length > 0 ? named.join(" · ") : null
}

/** Slower than the aircraft poll: a ship under way covers a few hundred metres
 *  in a minute, and asking faster is not more truthful, only more expensive for
 *  a public authority's server. */
export const VESSEL_POLL_MS = 45_000

/** The rail rows, in the order a reader looks for them. Ordered by how much of
 *  the water they are: trade first, then the boats that serve it. */
export const VESSEL_CATEGORIES: { key: VesselCategory; label: string; hint: string }[] = [
  { key: "cargo", label: "Cargo", hint: "freight and container traffic" },
  { key: "tanker", label: "Tankers", hint: "oil, chemical and gas carriers" },
  { key: "passenger", label: "Passenger", hint: "ferries, cruise and high-speed craft" },
  { key: "fishing", label: "Fishing", hint: "vessels declaring themselves fishing" },
  { key: "pleasure", label: "Pleasure", hint: "sailing and private craft" },
  { key: "service", label: "Service", hint: "tugs, pilots, rescue, dredgers" },
  { key: "other", label: "Unclassified", hint: "no type transmitted" },
]

/**
 * What is wrong with a position, in words a reader can act on.
 *
 * The mark is drawn either way. A transmitter claiming to be a ship in the
 * middle of a forest is a real thing that is really happening and is worth
 * more than the traffic around it — but it must not be drawn as an ordinary
 * vessel, so it says which test it failed.
 */
export function suspectReason(v: PresenceVessel): string | null {
  switch (v.position_suspect) {
    case "speed":
      return "reported speed no vessel can reach"
    case "stacked":
      return "several moving vessels on one position"
    default:
      return null
  }
}

/** The name a listener would hear first, falling back the way AIS degrades:
 *  the ship's name, then its call sign, then the number its transponder always
 *  sends. */
export function vesselTitle(v: PresenceVessel): string {
  const name = v.name?.trim()
  if (name) return name
  const callsign = v.callsign?.trim()
  if (callsign) return callsign
  if (v.mmsi != null) return `MMSI ${v.mmsi}`
  return "Unknown vessel"
}

/**
 * Whether the mark should point anywhere.
 *
 * Heading is what the hull is pointing at and course is where it is going;
 * they differ in a current, and heading is the truer one for drawing a hull.
 * A vessel that sent neither is drawn unrotated, because north would be a
 * direction it never claimed.
 */
export function vesselBearing(v: PresenceVessel): number | null {
  if (v.heading != null) return v.heading
  if (v.course != null) return v.course
  return null
}

/** Whether a vessel is going anywhere. Below half a knot is noise in a moored
 *  ship's own position, not movement, and drawing it as under way would put a
 *  wake on a hull that has been tied up for a week. */
export function vesselIsUnderWay(v: PresenceVessel): boolean {
  if (v.nav_status === "at anchor" || v.nav_status === "moored" || v.nav_status === "aground") {
    return false
  }
  return (v.speed_kt ?? 0) >= 0.5
}

/**
 * What the vessel is currently transmitting, as label/value lines.
 *
 * Absent fields are left out rather than rendered blank, the same rule the
 * aircraft card follows: an empty value reads as a measurement of nothing when
 * what happened is that nothing was measured.
 */
export function vesselFacts(v: PresenceVessel): { label: string; value: string }[] {
  const out: { label: string; value: string }[] = []
  const category = VESSEL_CATEGORIES.find((c) => c.key === v.category)
  if (category && v.category !== "other") out.push({ label: "type", value: category.label })
  if (v.nav_status) out.push({ label: "status", value: v.nav_status })
  if (v.speed_kt != null) {
    out.push({
      label: "speed",
      value: v.speed_kt < 0.5 ? "stopped" : `${v.speed_kt.toFixed(1)} kt`,
    })
  }
  const bearing = vesselBearing(v)
  if (bearing != null) out.push({ label: "heading", value: `${Math.round(bearing)}°` })
  const destination = v.destination?.trim()
  //: What the crew typed, and nothing more. It is a free-text field on a form,
  //: it is often stale and sometimes a joke, and the card says whose claim it
  //: is rather than presenting it as a fact about where the ship will go.
  if (destination) out.push({ label: "says bound for", value: destination })
  const callsign = v.callsign?.trim()
  if (callsign) out.push({ label: "call sign", value: callsign })
  if (v.imo != null) out.push({ label: "IMO", value: String(v.imo) })
  if (v.mmsi != null) out.push({ label: "MMSI", value: String(v.mmsi) })
  return out
}
